### PR TITLE
Anfix

### PR DESCRIPTION
**Pull request by michaelcfanning**
This change provides a bulk fix capability of the built-in CS8019 (unused namespace import) analysis.

### CODE REVIEW

**Comment by michaelcfanning** [comment]
@lgolding @Priya91 @jaredpar

File: src/Microsoft.DotNet.CodeFormatting.Tests/Analyzers/OptimizeNamespaceImportsTests.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+public class Test
+{
+    public static void Main()
+    {
+        Console.WriteLine(""Calling Console.WriteLine, a dependency on System namespace."");
+    }
+}
+";
+            Verify(input, expected, runFormatter: false);
+        }
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting.Tests/Analyzers/OptimizeNamespaceImportsTests.cs]: Blank line after closing braces.

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DotNet.CodeFormatting.Analyzers
+{
+    [Export(typeof(DiagnosticAnalyzer))]
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class OptimizeNamespaceImportsAnalyzer : DiagnosticAnalyzer
```
Comment (jaredpar) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: Could be sealed

**Comment by jaredpar** [comment]
:+1:

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs (hunk 2 above)
Comment (michaelcfanning) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: yes, will change. what do you think about this auto-fix 'automatically seal unextended classes'? :)

Comment (jaredpar) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: @michaelcfanning 

Unfortunately can't be done with `public` types :(

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                                                                            true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(s_rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSemanticModelAction(semanticModelAnalysisContext =>
+            {
+                SyntaxNode root;
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: Let us let @jaredpar be the authority on whether he wants to combine-declaration-with-initialization in this code base. You know how I feel about it ;-)

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                SemanticModel semanticModel;
+
+                semanticModel = semanticModelAnalysisContext.SemanticModel;
+                root = semanticModel.SyntaxTree.GetRoot();
+
+                // If we encounter any conditionally included code, we cannot be sure
+                // that unused namespaces might not be relevant for some other compilation
+                if (root.DescendantTrivia().Any(x => x.Kind() == SyntaxKind.IfDirectiveTrivia))
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: Likewise whether @jaredpar wants braces around single-line conditionals.

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                    return;
+
+                var diagnostics = semanticModel.GetDiagnostics(null, semanticModelAnalysisContext.CancellationToken);
+                Diagnostic firstDiagnostic = null;
+                var locations = new List<Location>();
+
+                foreach (Diagnostic diagnostic in diagnostics)
+                {
+                    if (diagnostic.Id == "CS8019")
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: `private const string UnnecessaryUsingDirective = "CS8019";`

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    {
+                        firstDiagnostic = firstDiagnostic ?? diagnostic;
+                        locations.Add(diagnostic.Location);
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsAnalyzer.cs]: As the code stands, `locations` combines all the locations, including the first one. But I don't think the `additionalLocations` parameter to `Diagnostic.Create` is supposed tyo include the first one.

File: src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsFixer.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class OptimizeNamespaceImportsFixer : CodeFixProvider
+    {
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            Diagnostic diagnostic = context.Diagnostics.First();
+            var usingDirectiveNodes = new List<SyntaxNode>();
+
+            // We recapitulate the primary diagnostic location in the
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Analyzers/OptimizeNamespaceImportsFixer.cs]: OK, this addresses my comment above as far as why you did it. Sri says that it was intended just to have the "others", but that is not documented, and Roslyn itself doesn't rely on that behavior anywhere. So I guess this is fine.

File: src/Microsoft.DotNet.CodeFormatting/Resources.resx [hunk 8]
```diff
@@ -126,6 +126,15 @@
   <data name="ExplicitThisFixer_Title" xml:space="preserve">
     <value>Remove 'this' qualifier</value>
   </data>
+  <data name="OptimizeNamespaceImportsAnalyzer_MessageFormat" xml:space="preserve">
```
Comment (lgolding) [inline src/Microsoft.DotNet.CodeFormatting/Resources.resx]: Suggest calling the analyzer itself `UnusedNamespaceDirectivesAnalyzer`. Just my 2c.

Comment (srivatsn) [inline src/Microsoft.DotNet.CodeFormatting/Resources.resx]: +1

**Pull request merged**